pkg/vms/actions.es6

```diff
@@ -0,0 +1,179 @@
+ *
+ * Lazily initializes the virt provider and dispatches given method on it.
+ */
+function virt (method, action) {
+  return (dispatch, getState) => getVirtProvider({ dispatch, getState }).then(provider => {
+    if (method in provider) {
+      console.log(`Calling ${provider.name}.${method}(${JSON.stringify(action)})`);
+      return dispatch(provider[method](action));
+    } else {
+      console.warn(`method: '${method}' is not supported by provider: '${provider.name}'`);
+    }
+  }).catch(err => {
+    console.error('could not detect any virt provider');
+  })
+}
+
+function getVirtProvider(store) {
+  const state = store.getState();
+  if (state.config.provider) {
+    return cockpit.resolve(state.config.provider);
+  } else {
+    const deferred = cockpit.defer();
+    console.log('Discovering provider');
+    /* TODO: discover host capabilities
+     systemctl is-active vdsmd
+       active
+       unknown
+     */
+    let provider = null;
+    if (false /*TODO: Detect VDSM*/) {
+      // TODO: dispatch/resolve VDSM provider
+    } else if (true /* TODO: detect libvirt */) {
+      console.log('Selecting Libvirt as the VIRT provider.');
+      provider = Libvirt;
+    }
+
+    if (!provider) { //  no provider available
+      deferred.reject();
+    } else {
+      store.dispatch(setProvider(provider));
+
+      // Skip the initialization if provider does not define the `init` hook.
+      if (!provider.init) {
+        deferred.resolve(provider);
+      } else {
+        // Providers are expected to return promise as a part of initialization
+        // so we can resolve only after the provider had time to properly initialize.
+        store
+          .dispatch(provider.init())
+          .then(() => deferred.resolve(provider))
+          .catch(deferred.reject);
+      }
+    }
+
+    return deferred.promise;
+  }
+}
+
+/**
+ * Helper for delaying the execution of requested action
+ */
+export function delay (action, timeout) {
+  return (dispatch, getState) => {
+    timeout = timeout || getRefreshInterval(getState());
+    if (timeout > 0) {
+      console.log(`Scheduling ${timeout} ms delayed action`);
+      window.setTimeout(() => dispatch(action), timeout);
+    } else {
+      console.log(`Skipping delayed action since refreshing is switched off`);
+      window.setTimeout(() => dispatch(delay(action)), VMS_CONFIG.DefaultRefreshInterval);
+    }
+  }
+}
+
+// --- Store actions --------------------------------------------
+export function setProvider (provider) {
+  return {
+    type: 'SET_PROVIDER',
+    provider
+  }
+}
+
+export function setRefreshInterval (refreshInterval) {
+  return {
+    type: 'SET_REFRESH_INTERVAL',
+    refreshInterval
+  }
+}
+
+export function clearVms (vms) {
+  return {
+    type: 'CLEAR_VMS'
+  }
+}
+
+export function updateOrAddVm ({ id, name, state, osType, fqdn, uptime, currentMemory, vcpus, autostart }) {
+  let vm = {};
+  if (id !== undefined) vm.id = id;
```

Comment:
Shouldn't this use an Object.assign() or similar function? Or should callers just use Object.assign()?